pkg/report/github/github.go

```diff
@@ -105,8 +105,17 @@ func (w Writer) Write(ctx context.Context, report types.Report) error {
 		manifest.Name = string(result.Type)
 		// show path for language-specific packages only
 		if result.Class == types.ClassLangPkg {
-			manifest.File = &File{
-				SrcLocation: result.Target,
```

Comment:
We aggregate some packages to 1 result - https://github.com/aquasecurity/trivy/blob/fb36c4ed09efc3fc241d02713c4cc864b6c6a2c8/pkg/fanal/applier/docker.go#L263-L291

But we can find path from Results.Packages.FilePath:
```
{
      "Target": "Python",
      "Class": "lang-pkgs",
      "Type": "python-pkg",
      "Packages": [
        {
          "Name": "pip",
          "Version": "23.2.1",
          "Licenses": [
            "MIT"
          ],
          "Layer": {
            "DiffID": "sha256:a7d483b6066e2ed6c9cd1d5890a93326e3f192ef6e537c046296053fa547a844"
          },
          "FilePath": "usr/local/lib/python3.12/site-packages/pip-23.2.1.dist-info/METADATA"
        },
```

Maybe we just use FilePath for these packages?

Reply:
Although I agree with you it's an interesting change as it would make the manifest more precise.

Nonetheless, I think the full `FilePath` would be too much to render in Github Dependency and would bloat the UI.
I can do a test with that change if you want and show the results here ?

Otherwise we could try to extract important values from the `FilePath` and adding them to `result.Target`, but that seems complex and prone to errors.

Reply:
So I have a few things to report:

1. As you pointed out trivy aggregates multiple packages in the same "App" (for instance `Python` or `Node`) into one result.
The issue is that in the sbom format, the `source_location` is set at the `manifest` level not at the package level (`resolved` key in sbom format). Meaning, if we keep the default trivy aggregation then we have to use only one of all the packages `FilePath`. How to define that value is either arbitrary, or we need some sort of heuristic ?

2. In terms of UI it doesn't as bloated as I expected so I think you're right and it's ok to use.
![Screenshot_20240126_150104](https://github.com/aquasecurity/trivy/assets/72691393/5e4c37a3-bdbb-46c2-8d3b-1a06ea99566b)
The `source_location` used in the above screenshot is simply the `FilePath` of the last package in `result.Packages` so it's not even correct.


3. In the case trivy is scanning an image I think it's still pretty useless to replace the `manifest`'s `source_location` with this `FilePath` value and we should instead reference the image's name and tag.

Reply:
> Meaning, if we keep the default trivy aggregation then we have to use only one of all the packages FilePath. How to define that value is either arbitrary, or we need some sort of heuristic ?

We have already seen cases where it is necessary to remove aggregation, but this requires large changes and we are not sure about it.
I created a small function to get filePath for each package - https://github.com/DmitriyLewen/trivy/commit/84316646c3c4d75f6b79fc885e5a93811b87525e
Can you take a look?

> In the case trivy is scanning an image I think it's still pretty useless to replace the manifest's source_location with this FilePath value and we should instead reference the image's name and tag.

This makes sense for UI. But what worries me is case where user only knows that image contains vulnerable package, but doesn't have filePath to find and update that package. It would be great to add filePath somehow.
e.g. we can use `<image_name>:<tag> - <filePath>` format for [manifest map key](https://github.com/aquasecurity/trivy/blob/4df9363890ca7268d5c262a5916da0374921148b/pkg/report/github/github.go#L130). Is it possible to see `gsbom` file after uploading to GitHub?

Reply:
> I created a small function to get filePath for each package - https://github.com/DmitriyLewen/trivy/commit/84316646c3c4d75f6b79fc885e5a93811b87525e
Can you take a look?

Just had a look, so from the trivy `Results` there is still aggregation but the sbom `Manifest` is not keeping this aggregation. Meaning each package is its own result with its own `source_location`, that seems like a good compromise to me.

> But what worries me is case where user only knows that image contains vulnerable package, but doesn't have filePath to find and update that package.

My assumption is that most people scanning images are in charge of maintaining those, meaning they know what's installed on them and how, so knowing the exact `FilePath` of a dependency won't change much.

But it's true we could always do `<image_name>:<tag> - <filePath>` although it's going to be super long and I'm not sure how github manifest filter will react to it.

I feel like maybe we should ask a broader audience then just the 2 of us for the direction to go, but unsure how and where to ask.

> Is it possible to see gsbom file after uploading to GitHub?

Yes. From my testing it's not exactly the same SBOM file trivy sends to github, but [you can still export your SBOM report from github. ](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/exporting-a-software-bill-of-materials-for-your-repository#exporting-a-software-bill-of-material-for-your-repository-from-the-ui)

Reply:
Hello @knqyf263 
Can you take part in this discussion?
Perhaps you'll see another idea.

Reply:
See https://github.com/aquasecurity/trivy/pull/5999#discussion_r1476156309

I think the drawback of https://github.com/DmitriyLewen/trivy/commit/84316646c3c4d75f6b79fc885e5a93811b87525e (each package becoming it's own manifest) would ruin the basic usage of Github Dependency as it would make filtering by manifest very complex.

I went ahead and pushed the change to use `<image_registry>/<image_name>:<tag>@sha256:<image_hash>`, until we find a better solution.

Reply:
>My assumption is that most people scanning images are in charge of maintaining those, meaning they know what's installed on them and how, so knowing the exact FilePath of a dependency won't change much.

Unfortunately, it is not true. We've been seeing a lot of issues with people being confused about where a package is installed. This is because our table format doesn't include the full path. Therefore, we always point out that the package path is included in the JSON report and that they should check the path.

I might not be following the conversation., but what if adding file paths in `metadata`? 
https://github.com/aquasecurity/trivy/blob/b65ef00bca9e25a558222868f26de0641e9341e3/pkg/report/github/github.go#L160-L169

Reply:
Do you mean to get format like https://github.com/aquasecurity/trivy/pull/5999#discussion_r1476156309?

Reply:
IIUC, it just puts a file path to a key. I'm suggesting we put file paths under the metadata of packages. BTW, the line I shared above was wrong.
https://github.com/aquasecurity/trivy/blob/b65ef00bca9e25a558222868f26de0641e9341e3/pkg/report/github/github.go#L32

Or I may not understand the problem correctly. I just answered this question.

>But what worries me is case where user only knows that image contains vulnerable package, but doesn't have filePath to find and update that package. It would be great to add filePath somehow.

https://docs.github.com/en/rest/dependency-graph/dependency-submission?apiVersion=2022-11-28

Reply:
I just tested and adding a `metadata` at the package level package:
```
"Pillow": {
  "package_url": "pkg:pypi/pillow@10.1.0",
  "relationship": "direct",
  "scope": "runtime",
  "metadata": {
    "source_location": "usr/local/lib/python3.12/site-packages/pillow-10.1.0.dist-info/METADATA"
  }
},
```
doesn't change anything in Github Dependency UI.

But I agree at least the user would be able to find the path for each package by parsing the JSON.

Reply:
Interesting. Isn't the metadata shown anywhere?

Reply:
Not that I can see in Github Dependency UI no, see screenshot below to see how it looks.

![Screenshot_20240208_065410](https://github.com/aquasecurity/trivy/assets/72691393/99749d6c-bc60-422c-8d3f-85a467027daa)

Reply:
So user can only get package metadata when [taking snapshot](https://docs.github.com/en/rest/dependency-graph/dependency-submission?apiVersion=2022-11-28#create-a-snapshot-of-dependencies-for-a-repository), right?
So the user can only get the package metadata when taking a snapshot, right?

Reply:
If they use trivy then there's no issue as the user will now be able to find the file path for each package by parsing the JSON report before sending it to GitHub.

But if they use trivy-action then the report is sent automatically and the user doesn't have access to the original report hence he wouldn't be able to parse it.
Maybe if the user then export it from GitHub, the metadata tag will stay, I'll test that.

Reply:
This is what I got from github export to SBOM for the same package as in https://github.com/aquasecurity/trivy/pull/5999#discussion_r1479587202:
```
{
      "SPDXID": "SPDXRef-pip-pillow-10.1.0",
      "name": "pip:pillow",
      "versionInfo": "10.1.0",
      "downloadLocation": "NOASSERTION",
      "filesAnalyzed": false,
      "licenseConcluded": "CAL-1.0",
      "supplier": "NOASSERTION",
      "externalRefs": [
        {
          "referenceCategory": "PACKAGE-MANAGER",
          "referenceLocator": "pkg:pypi/pillow@10.1.0",
          "referenceType": "purl"
        }
      ]
    },
```

As you can see the `metadata` tag doesn't show up the the exported SBOM.

> [!IMPORTANT] 
> Conclusion: users of `trivy-action` would currently have no way of finding out the `filePath` of each package when using the `github` format.

As this issue only involves `trivy-action` and not `trivy` itself anymore, I think we  should continue this discussion to find a solution for `trivy-action` users at https://github.com/aquasecurity/trivy-action/issues/286
(I would recommend to add an option for [upload as an artifact](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts) so that user can easily fetch the report).

Reply:
I created https://github.com/aquasecurity/trivy-action/pull/307 to make sure users of `trivy-action` will know where to find the missing details from Github Dependency.

@DmitriyLewen and @knqyf263 do you agree with this compromise ?
If so, can you review this PR code ?